synapse/storage/databases/main/event_push_actions.py

```diff
@@ -436,42 +454,70 @@ def _get_unread_counts_by_pos_txn(
             """,
             (room_id, user_id, receipt_stream_ordering, receipt_stream_ordering),
         )
-        row = txn.fetchone()
+        max_summary_stream_ordering = 0
+        for summary_stream_ordering, notif_count, unread_count, thread_id in txn:
+            if thread_id == "main":
+                counts = NotifCounts(
+                    notify_count=notif_count, unread_count=unread_count
+                )
+            # TODO Delete zeroed out threads completely from the database.
+            elif notif_count or unread_count:
+                thread_counts[thread_id] = NotifCounts(
+                    notify_count=notif_count, unread_count=unread_count
+                )
 
-        summary_stream_ordering = 0
-        if row:
-            summary_stream_ordering = row[0]
-            counts.notify_count += row[1]
-            counts.unread_count += row[2]
+            # XXX All threads should have the same stream ordering?
+            max_summary_stream_ordering = max(
+                summary_stream_ordering, max_summary_stream_ordering
+            )
```

Comment:
There's a thread about this at https://github.com/matrix-org/synapse/pull/13181#discussion_r955192602, need to investigate.

Reply:
@erikjohnston said:

> I think that stream ordering is the "max" stream ordering that we rotated from the event_push_actions table:
>
> https://github.com/matrix-org/synapse/pull/13181/files#diff-f121377d76a7b35c60092da2c4bf8c849544459a2c676cf8e87057aff24ece54R1218
>
> I think that means its equivalent to the stream ordering we rotated up to? They may differ but there should never be any relevant rows in EPA between the per-thread stream ordering and the max stream ordering?

To reword this a bit: for each `(user, room, thread)` tuple we store the latest stream ordering *in that thread* into the summary. So what @erikjohnston is saying is that this is pretty much equivalent to using the value stored in `event_push_summary_stream_ordering`, I think.

I'll try that out on `develop` (sans threading) first and see if that tracks.

Reply:
> this is pretty much equivalent to using the value stored in `event_push_summary_stream_ordering`, I think.

This isn't quite accurate though because receipts can be earlier than the latest summary (and some other edge-cases). Pretty much the `stream_ordering` is also used to let us know if the summary is valid.

Reply:
I clarified this comment in cb679e2379f981bc4fa3675b9ac37d642dc32985 (and removed the `XXX`). I think this code is valid, I just was not quite sure *why* it was valid. 👍